egl-wayland: Read compositor device from zwp_linux_dmabuf_v1

Molytho opened the pull request:
Compositors start dropping support for wl_drm because it is replaced by linux-dmabuf. See https://gitlab.freedesktop.org/wlroots/wlroots/-/commit/2f2a55ebd3c5a950f1bf8c280edf0838721363da. Try reading compositors main device from linux-dmabuf v4 if supported and otherwise fallback to wl_drm.

erik-kz commented:
Thanks for the patch! Note, however, that we already bind the `zwp_linux_dmabuf_v1` protocol and retrieve the default feedback during `wlEglInitializeHook`. So we shouldn't need to do it a second time.

I *do* think it's a good idea to use `zwp_linux_dmabuf_v1` to determine the compositor device instead of `wl_drm` when possible. If you could modify the patch to keep that part but use the existing `WlEglDisplay::defaultFeedback` I would be happy to merge it.

Molytho commented:
I changed the code to bind to zwp_linux_dmabuf_v1 only once.
I also added some error handling to getServerProtocolsInfo.
Sadly this made it more ugly

erik-kz commented on src/wayland-egldisplay.c:
```diff
@@ -435,8 +436,9 @@ registry_handle_global(void *data,
         /*
          * Version 3 added format modifier support, which the dmabuf
          * support in this library relies on.
+         * If the version is >= 4 we already bound it in getServerProtocolsInfo
```
I think it might simplify thing a little bit to always bind `zwp_linux_dmabuf_v1` during `eglGetPlatformDisplay`, regardless of whether version 3 or 4 is supported. Of course, if version 4 isn't supported we also need to bind `wl_drm`. Then we can get rid of the code that binds it during `eglInitialize`.

Molytho replied:
Sounds good. This sadly still leaves on issue:
if the display is initialized and then destroyed we also free all wayland objects in terminateDisplay. The spec allows initializing the same display again so we would still need to bind the interfaces in eglInitialize if they aren't bound already.

erik-kz replied:
Ah shit, true, I had forgotten about that. Well, if we might need to re-bind the interface during `eglInitialize` anyway maybe it would be better to just go with the original version of your patch. Now I feel like a bit of an ass for making you re-write it, really sorry about that.

Molytho replied:
No problem. I initially also forgot about that case.
I reverted to the initial patch but kept the new error handling in getServerProtocolsInfo

erik-kz commented on src/wayland-egldisplay.c:
```diff
@@ -670,48 +674,103 @@ EGLBoolean wlEglTerminateHook(EGLDisplay dpy)
+    if (!wlRegistry) {
+        goto out;
+    }
     wl_proxy_wrapper_destroy(wrapper); /* Done with wrapper */
```
I think we can remove this (and the line below) since we're now destroying the wrapper at the end.

erik-kz commented on src/wayland-egldisplay.c:
```diff
@@ -898,7 +960,7 @@ EGLDisplay wlEglGetPlatformDisplayExport(void *data,
     }
 
-    if (!protocols.hasDrm || (!protocols.hasEglStream && !protocols.hasDmaBuf)) {
+    if (!protocols.hasEglStream && !protocols.hasDmaBuf) {
```
Do we really need to change this condition? If we don't have `zwp_linux_dmabuf_v1`  or `wl_eglstream` but we do have `wl_drm` I think we're still fine, are we not?

Molytho replied:
We definitely have to change this because this always results in a failure if wl_drm is not available.
I don't think that we are fine with wl_drm alone. This if statement currently leads to a fail if neither zwp_linux_dmabuf_v1 nor wl_eglstream are available

erik-kz replied:
Oh, yeah you're right. Sorry, I should have thought it through more carefully.

erik-kz commented on src/wayland-egldisplay.c:
```diff
@@ -670,48 +674,103 @@ EGLBoolean wlEglTerminateHook(EGLDisplay dpy)
+        if (drm_device->available_nodes & (1 << DRM_NODE_RENDER)) {
+            result = strdup(drm_device->nodes[DRM_NODE_RENDER]);
+        } else if (drm_device->available_nodes & (1 << DRM_NODE_PRIMARY)) {
+            result = strdup(drm_device->nodes[DRM_NODE_PRIMARY]);
```
In `wlEglGetPlatformDisplayExport` we only check if the render node of an `EGLDevice` matches `drm_name`. Therefore I don't think there's any point returning a primary node path from this function. If there's no render node available we can just return NULL.

erik-kz commented:
Thanks. This looks good to me. I just need to get a +1 from another NVIDIA engineer and then I will merge it.

cubanismo commented on src/wayland-egldisplay.c:
```diff
@@ -500,6 +502,95 @@ static const struct wl_drm_listener drmListener = {
+
+    assert(protocols->devId != 0);
+    drmDevice *drm_device;
+    if (drmGetDeviceFromDevId(protocols->devId, 0, &drm_device) == 0) {
```
This is a relatively new function in libdrm. I don't see a particular version of libdrm required in the  current meson.build. Perhaps we should add a requirement for version 2.4.109, the first version to include this function?

kbrenneman replied:
Are there any systems that are both new enough to support Wayland and old enough to not have `drmGetDeviceFromDevId`?

If so, then we'd probably need to look up `drmGetDeviceFromDevId` at runtime using dlsym.

erik-kz replied:
2.4.109 was released May of 2021, so maybe? But if we're going to do that would it be easier to just return `"/dev/dri/renderD" + minor(protocols->devId)`? It looks like that's pretty much how libdrm populates `drmDevice::nodes`.

cubanismo replied:
Even though I think that's reliable in practice, I don't think we're supposed to rely on file layout assumptions. It probably is safe to assume systems without DRM 2.4.109 are running compositors that still support wl_drm though, so a dynamic lookup and skipping this path if it isn't there seems fine. Normally I'd say just compile it out if the DRM version is too old, but since we'll have to ship this as a binary in our driver package as well, I agree a runtime check is needed.

Molytho replied:
I updated the code to use dlsym now

cubanismo commented on src/wayland-egldisplay.c:
```diff
@@ -670,48 +763,74 @@ EGLBoolean wlEglTerminateHook(EGLDisplay dpy)
-            /* destroy our wl_drm object */
-            wl_drm_destroy(protocols->wldrm);
+        }
+        result = protocols->drm_name != NULL;
```
I see some discussion of this, but is this really fatal? I thought the streams path worked fine without this.

erik-kz replied:
I think it kinda doesn't matter since `checkNvidiaDrmDevice` will return false if `protocols->drm_name` is NULL, meaning we will fail `eglGetPlatformDisplay` anyway unless `__NV_PRIME_RENDER_OFFLOAD` is set. But PRIME render-offload definitely doesn't doesn't work with the EGLStreams path.

cubanismo replied:
Hrm. Did we regress that as part of the not-waking-up-GPUs-for-every-eglGetDisplay() change? It seems like if so, we sort of broke the eglstreams path. IIRC the wl_drm usage was added for the dma-buf path, so I doubt the eglstreams path was intended to require it.

erik-kz replied:
The EGLStreams path only works if the compositor calls `eglBindWaylandDisplayWL`, since that's where we bind the `wl_eglstream` interface. But in that case we will also bind our own implementation of `wl_drm`. So we shouldn't really need to worry about the EGLStreams path working without `wl_drm`, it will always be available.

To your question, though, no I don't think the change you mentioned regressed anything. All it did was move the failure point a little bit earlier. But even before that change, if we couldn't determine what device the compositor was using we would  fail (unless `__NV_PRIME_RENDER_OFFLOAD` is set).

cubanismo replied:
OK. perhaps I'm misremembering.

erik-kz replied:
I mean, there *was* a time when we would just choose a device completely arbitrarily, with no concern for which device the compositor was using. But we needed to change that because it was problematic on PRIME systems. See https://github.com/NVIDIA/egl-wayland/commit/d4937adc5cd04ac7df98fc5616e40319fb52fdee

cubanismo reviewed (commented):
A suggestion and a question inline.

cubanismo commented on src/wayland-egldisplay.c:
```diff
@@ -670,48 +768,79 @@ EGLBoolean wlEglTerminateHook(EGLDisplay dpy)
         wl_display_roundtrip_queue(nativeDpy, queue);
-        if (protocols->hasDrm) {
+        if (!getDeviceFromDevIdInitialised) {
+            getDeviceFromDevId = dlsym(RTLD_DEFAULT, "drmGetDeviceFromDevId");
```
Thanks for adding this, but I think you also need to update the meson and automake/autoconf files since this is the first use of libdl in egl-wayland AFAIK. For an example of the meson changes, see the egl-gbm equivalent:

https://github.com/NVIDIA/egl-gbm/blob/b7d9f41a959c00d83d948ec1a17a287f3ce7f9aa/src/meson.build#L1

I'm not familiar enough with autoconf myself to make a suggestion there.

erik-kz replied:
We actually used to link against libdl. You could just revert the changes to Makefile.am, configure.ac, and Meson.build made by  https://github.com/NVIDIA/egl-wayland/commit/2eb4628d64a8297fb4c08aa796fdf33ff54a670f

Molytho replied:
Thanks. This was quiet helpful. Should be done now.

cubanismo replied:
Awesome, thanks.

cubanismo reviewed (approved):
Latest version LGTM.
